google-cloud-pubsub/src/main/java/com/google/cloud/pubsub/v1/MessageDispatcher.java

```diff
@@ -372,13 +386,48 @@ void processReceivedMessages(List<ReceivedMessage> messages) {
         // totally expire so that pubsub service sends us the message again.
         continue;
       }
-      outstandingBatch.add(new OutstandingMessage(message, ackHandler));
+      OutstandingMessage outstandingMessage = new OutstandingMessage(message, ackHandler);
+      if (this.exactlyOnceDeliveryEnabled.get()) {
+        exactlyOncePendingBatch.add(outstandingMessage);
+        outstandingReceipts.put(message.getAckId(), outstandingMessage);
+      } else {
+        outstandingBatch.add(outstandingMessage);
+      }
       pendingReceipts.add(ackRequestData);
     }
-
     processBatch(outstandingBatch);
   }
 
+  void notifyAckSuccess(AckRequestData ackRequestData) {
+
+    if(outstandingReceipts.containsKey(ackRequestData.getAckId())) {
+      OutstandingMessage outstandingMessage = outstandingReceipts.get(ackRequestData.getAckId());
+      pendingMessages.putIfAbsent(outstandingMessage.receivedMessage.getAckId(),
+          outstandingMessage.ackHandler);
+
+      exactlyOnceOutstandingBatch.add(outstandingMessage);
```

Comment:
I worry with 3 additional data structures that all depend on each other, this is going to get confusing and we are going to forget to remove from one of the maps at some point. We should be able to simplify logic here a bit, and just have 1 additional data structure:
1) outstandingReceipts - LinkedHashMap<String, <bool, OutstandingMessage>> - a linked map that tracks insertion order. Key is the ackId, and value is a tuple of the outstandingMessage and a bool that tracks whether or not the receipt has completed or not.

on notifyAckSuccess, we should just have to:
1) check if the ackId is in outstandingReceipts, and set the corresponding bool to true to indicate receipt completed.
2) Then we can easily just keep popping from the outstandingReceipts list until we hit a receipt that hasn't completed yet.
 

When notified of ack success, we should just add to completedReceipts